tsfc/driver.py

```diff
@@ -491,6 +377,60 @@ def compile_expression_dual_evaluation(expression, to_element, *,
     return builder.construct_kernel(return_arg, impero_c, index_names, first_coefficient_fake_coords)
 
 
+class DualEvaluationCallable(object):
+    """
+    Callable representing a function to dual evaluate.
+
+    When called, this takes in a
+    :class:`finat.point_set.AbstractPointSet` and returns a GEM
+    expression for evaluation of the function at those points.
+
+    :param expression: UFL expression for the function to dual evaluate.
+    :param kernel_cfg: A kernel configuration for creation of a
+        :class:`GemPointContext` or a :class:`PointSetContext
+
+    Not intended for use outside of
+    ``compile_expression_dual_evaluation``.
+    """
+    def __init__(self, expression, kernel_cfg):
+        self.expression = expression
+        self.kernel_cfg = kernel_cfg
+        # TODO: Deal with case when expression can be split into subexpressions
+        # for dual evaluation on FInAT tensor product elements. If so add a
+        # ``self.factors`` property which can be tested for existence of in
+        # FInAT's tensor product element dual basis method.
```

Comment:
How would you test for such a structure so you can dispatch to the correct dual basis building routine? Here's what's in the linked FInAT branch https://github.com/FInAT/FInAT/blob/f3eff94f05db3e73cd63958df13ef0f17aa9f355/finat/tensor_product.py#L182-L223

Reply:
So the things you need to optimise are:
1. the contraction that evaluates at points (I think this will happen automagically right now)
2. the contraction of `Q` against `fn(x)`, it's possible that if `Q` and `fn(x)` have the right gem sub-structure then `gem.optimise.contraction` will do that for you (without a need for `factors` anywhere).

Reply:
So you reckon replacing https://github.com/FInAT/FInAT/blob/f3eff94f05db3e73cd63958df13ef0f17aa9f355/finat/tensor_product.py#L207 with `gem.optimise.contraction` should just DTRT and there's no need for any fancy dispatching? How would one test if that is true?